qa/L0_backend_python/response_sender/response_sender_test.py

```diff
@@ -0,0 +1,293 @@
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import time
+import unittest
+
+import numpy as np
+import tritonclient.grpc as grpcclient
+
+
+class ResponseSenderTest(unittest.TestCase):
+    def _get_inputs(
+        self,
+        number_of_response_before_return,
+        send_complete_final_flag_before_return,
+        return_a_response,
+        number_of_response_after_return,
+        send_complete_final_flag_after_return,
+    ):
+        shape = [1, 1]
+        inputs = [
+            grpcclient.InferInput("NUMBER_OF_RESPONSE_BEFORE_RETURN", shape, "UINT8"),
+            grpcclient.InferInput(
+                "SEND_COMPLETE_FINAL_FLAG_BEFORE_RETURN", shape, "BOOL"
+            ),
+            grpcclient.InferInput("RETURN_A_RESPONSE", shape, "BOOL"),
+            grpcclient.InferInput("NUMBER_OF_RESPONSE_AFTER_RETURN", shape, "UINT8"),
+            grpcclient.InferInput(
+                "SEND_COMPLETE_FINAL_FLAG_AFTER_RETURN", shape, "BOOL"
+            ),
+        ]
+        inputs[0].set_data_from_numpy(
+            np.array([[number_of_response_before_return]], np.uint8)
+        )
+        inputs[1].set_data_from_numpy(
+            np.array([[send_complete_final_flag_before_return]], bool)
+        )
+        inputs[2].set_data_from_numpy(np.array([[return_a_response]], bool))
+        inputs[3].set_data_from_numpy(
+            np.array([[number_of_response_after_return]], np.uint8)
+        )
+        inputs[4].set_data_from_numpy(
+            np.array([[send_complete_final_flag_after_return]], bool)
+        )
+        return inputs
+
+    def _generate_streaming_callback_and_responses_pair(self):
+        responses = []  # [{"result": result, "error": error}, ...]
+
+        def callback(result, error):
+            responses.append({"result": result, "error": error})
+
+        return callback, responses
+
+    def _infer(
+        self,
+        model_name,
+        number_of_response_before_return,
+        send_complete_final_flag_before_return,
+        return_a_response,
+        number_of_response_after_return,
+        send_complete_final_flag_after_return,
+    ):
```

Comment:
Would it make sense to have a sanity check to catch user error here with something like this:

```python
    # Assert that one and only one of these values is True, as they are mutually exclusive
    self.assertEqual([send_complete_final_flag_before_return, return_a_response, send_complete_final_flag_after_return].count(True), 1)
```

Reply:
I intentionally left this open, because then we can test improper model behavior using this testing base, for example, sending a complete final flag and then returning a response.